src/Integrators.jl

```diff
@@ -1,53 +1,178 @@
+
+1. [H. Zhang and C. D. Batista, Phys. Rev. B 104, 104409 (2021)](https://arxiv.org/abs/2106.14125).
+2. [D. Dahlbom et al, Phys. Rev. B 106, 054423 (2022)](https://arxiv.org/abs/2204.07563).
 """
 mutable struct ImplicitMidpoint
     Δt   :: Float64
     atol :: Float64
 
+    # Averages the square of the dynamical drift term for use in
+    # `check_timestep`.
+    drift_squared :: OnlineStatistics{Float64}
+
     function ImplicitMidpoint(Δt; atol=1e-12)
         Δt <= 0 && error("Select positive Δt")
-        return new(Δt, atol)
+        return new(Δt, atol, OnlineStatistics{Float64}())
     end    
 end
 
 
+function reset_statistics(integrator::Union{Langevin, ImplicitMidpoint})
+    integrator.drift_squared = OnlineStatistics{Float64}()
+end
+
+"""
+    check_timestep(integrator; tol)
+
+Prints a suggested timestep scale for the relative error tolerance `tol`, based
+on statistics collected from previous time integration. A reasonable tolerance
+might be of order `tol = 1e-2`, but should be selected on a case-by-case basis.
+Note that `tol` is interpreted as the error for the deterministic part of a
+single integration timestep, which scales like `Δt²`. Because the stochastic
+Heun integration scheme is weakly convergent of order-1, errors in the estimates
+of averaged observables may instead scale like ` √tol ~ Δt`. It is the
+responsibility of the user to choose `tol` accordingly.
+
+The suggested timestep will be inversely proportional to the largest energy
+scale that plays a role in the dynamics. If there is a strong Langvin coupling
+``λ`` to the thermal bath, it will effectively rescale the energy scale.
+
+Timestep statistics stored in `integrator` will be reset after calling this
+function.
+"""
+function check_timestep(langevin::Langevin; tol, reset=true)
+    (; Δt, λ, kT) = langevin
+
+    drift_magnitude = sqrt(mean(langevin.drift_squared))
+    if isnan(drift_magnitude)
+        error("Run trajectory first to collect statistics")
+    end
+
+    # Heun integration is second-order accurate, so the local error from each
+    # deterministic timestep scales as dθ². Angular displacement per timestep dθ
+    # scales like dt |drift|, yielding err1 ~ (dt |drift|)^2
+    #
+    # The "relevant" error introduced by thermal noise should be quantified in
+    # terms of the effect on statistical observables. To avoid subtleties in
+    # defining this error, we instead naïvely assume it continues to be second
+    # order in `dt`. To determine the proportionality constant, consider the
+    # high-T limit, where each spin undergoes Brownian motion. Here, the
+    # diffusion constant D ~ λ kT sets an inverse time-scale. This implies err2
+    # ~ (dt λ kT)².
+    #
+    # The total error (err1 + err2) should be less than the target tolerance.
+    # After some algebra, this implies,
+    #
+    # dt ≲ sqrt(tol / (c₁² |drift|² + c₂² λ² kT²))
+    #
+    # for some empirical constants c₁ and c₂.
+
+    c1 = 1.0
+    c2 = 1.0
```

Comment:
Is it correct to assume that the appropriate constants are model-dependent?

Reply:
I think so. We are trying to find a mapping between error and dt, but this depends on the model and the quantity used to define "error". Since it's fundamentally ambiguous, I defaulted to a simple choice here, which seemed to work well for our examples.

The ratio between c2/c1 particularly hard to select. It seems natural that, when T gets "big enough", we want dt to scale like T at a fixed tolerance. But how should we define "big enough" relative to other energy scales in the system?